lib/util/documentvisibilitymonitor.js

```diff
@@ -0,0 +1,85 @@
+const EventEmitter = require('events').EventEmitter;
+
+/**
+ * @extends EventEmitter
+ * @emits documentVisiblePhase1
+ * @emits documentVisiblePhase2
+ */
+const supportedPhases = ['phase1', 'phase2'];
+/**
+ * exposes a mechanism for listening for document visibility in
+ * ordered phases.
+ * caller can subscriber for document visibility on phase1 and phase2
+ * phase1 callbacks will be executed first. Those callbacks may return async.
+ * phase2 callbacks will be executed only after phase1 callbacks are settled.
+ */
+class DocumentVisibilityMonitor {
```

Comment:
@makarandp0 I think this class can be simplified a great deal like this:
```js
'use strict';

/**
 * The {@link DocumentVisibilityMonitor} monitors the visibility state of the DOM
 * and executes the attached listeners in phase order when the DOM is visible.
 */
class DocumentVisibilityMonitor {
  /**
   * Constructor.
   * @param {number} [nPhases=1] - the number of phases
   */
  constructor(nPhases = 1) {
    Object.defineProperties(this, {
      _listeners: {
        value: []
      },
      _nPhases: {
        value: nPhases
      },
      _onVisibilityChange: {
        value: () => {
          if (document.visibilityState === 'visible') {
            this._emitVisible();
          }
        }
      }
    });

    for (let i = 0; i < nPhases; i++) {
      this._listeners.push(new Set());
    }
  }

  get _listenerCount() {
    return this._listeners.reduce((count, phaseListeners) => count + phaseListeners.size, 0);
  }

  /**
   * Call all the listeners. Makes sure that all listeners for a given phase
   * are executed before calling the listeners of the next phase.
   * @private
   */
  _emitVisible() {
    let promise = Promise.resolve();
    for (let phase = 1; phase <= this._nPhases; phase++) {
      promise = promise.then(() => this._emitVisiblePhase(phase));
    }
    return promise;
  }

  /**
   * Call all the listeners for a given phase.
   * @private
   */
  _emitVisiblePhase(phase) {
    const phaseListeners = Array.from(this._listeners[phase - 1]);
    return Promise.all(phaseListeners.map(listener => {
      const ret = listener();
      return ret instanceof Promise ? ret : Promise.resolve(ret);
    }));
  }

  /**
   * Start listening to the DOM visibility state change.
   * @private
   */
  _start() {
    document.addEventListener('visibilitychange', this._onVisibilityChange);
  }

  /**
   * Stop listening to the DOM visibility state change.
   * @private
   */
  _stop() {
    document.removeEventListener('visibilitychange', this._onVisibilityChange);
  }

  /**
   * Listen for the DOM to be visible at the given phase.
   * @param {number} phase
   * @param {function} listener
   * @returns {this}
   */
  onVisible(phase, listener) {
    this._listeners[phase - 1].add(listener);
    if (this._listenerCount === 1) {
      this._start();
    }
    return this;
  }

  /**
   * Stop listening for the DOM to be visible at the given phase.
   * @param {number} phase
   * @param {function} listener
   * @returns {this}
   */
  offVisible(phase, listener) {
    this._listeners[phase - 1].delete(listener);
    if (this._listenerCount === 0) {
      this._stop();
    }
    return this;
  }
}

module.exports = new DocumentVisibilityMonitor(2);
```

Reply:
@manjeshbhargav, I like your suggestion in general, but it does look like we are implementing some of eventEmitter here. also while running my unit tests against your suggestion, one immediate issue I found is the use of `Set` to store listeners - that breaks the case where same callback is registered multiple times. But I think overall this looks good, and since you have already taken effort to write this - I will fix the unit tests, and get it working.

Reply:
@makarandp0 We're only recreating a little bit of EventEmitter. Also, in your implementation, EventEmitter is only used to store listeners and nothing is emitted on the EventEmitter. So, we can just maintain the queue ourselves in a simpler fashion.